container/container_unix.go

```diff
@@ -171,26 +171,25 @@ func (container *Container) HasMountFor(path string) bool {
 	return exists
 }
 
-// UnmountIpcMounts uses the provided unmount function to unmount shm and mqueue if they were mounted
+// UnmountIpcMounts uses the provided unmount function to unmount shm if it was mounted
 func (container *Container) UnmountIpcMounts(unmount func(pth string) error) {
-	if container.HostConfig.IpcMode.IsContainer() || container.HostConfig.IpcMode.IsHost() {
+	if container.HasMountFor("/dev/shm") {
 		return
 	}
 
 	var warnings []string
 
-	if !container.HasMountFor("/dev/shm") {
-		shmPath, err := container.ShmResourcePath()
-		if err != nil {
-			logrus.Error(err)
-			warnings = append(warnings, err.Error())
-		} else if shmPath != "" {
-			if err := unmount(shmPath); err != nil && !os.IsNotExist(err) {
-				if mounted, mErr := mount.Mounted(shmPath); mounted || mErr != nil {
-					warnings = append(warnings, fmt.Sprintf("failed to umount %s: %v", shmPath, err))
-				}
+	// container.ShmPath should not be used here as it may point
+	// to the host's or other container's /dev/shm
+	shmPath, err := container.ShmResourcePath()
+	if err != nil {
+		logrus.Error(err)
+		warnings = append(warnings, err.Error())
```

Comment:
Yes, it is from the time when the code dealt with /dev/shm and /dev/mqueue, the latter was later dropped. In general, I try to resist the urge of rewriting the existing code, for the sake of commit brevity. But, indeed, this is the bug, as the message is printed twice, so I'll fix it.